File: String/Wildcard/Wildard.java
class Wardcard {
	public boolean isWildcardMatching(String s, String t) {
		if(s == null || t == null)
			return false;
		
		if(s.length() == 0 || t.length() == 0) {
			return false;
		}
		
		return internalMatching(s, 0, t, 0);
	}
	
	private boolean internalMatching(String s, int sLen, String t, int tLen) {
		if(slen == s.length() || tLen == t.length()) {
			if(slen == s.length() && tLen == t.length()) {
				return true;
			} else {
				return false;
			}
		}

		if(t.CharAt(tLen) == '*') {
			while(t.CharAt(tLen) == '*') {
				tLen++;
				
				if(tLen == t.length())
					return true;
			}
			
			while(sLen < s.length() && internalMatching(s, sLen, t, tLen)) {
				sLen++;
			}
			return sLen != s.length();
		} else if(s.CharAt(sLen) == t.CharAt(tLen) || t.CharAt(tLen) == '?') {
			internalMatching(s, sLen + 1, t, tLen + 1);
		} else 
			return false;
	}
}

File: String/Wildcard/Wildard.cc
class WildcardMatching{
	public:
		bool isWildcardMatching(string s, int sLen, String t, int tLen) {
			int start = 0, ss = 0, i = 0, j = 0;
			while(s[i]) {
				if(t[j] == '?' || s[i] == t[j]) {
					i++;
					j++
					continue;
				}
				if(t[j] == '*') {
					star = j++;
					ss = i;
					continue;
				}
				if(star) {
					j = star;
					i = ++ss;
					continue;
				}
				
				return false;
			}
			
			while(t[j] == '*')
				j++
			
			return !t[j];
		}
};